diesel/src/sqlite/connection/mod.rs

```diff
@@ -372,4 +372,105 @@ mod tests {
             .get_result::<(i32, i32, i32)>(&connection);
         assert_eq!(Ok((2, 3, 4)), added);
     }
+
+    use crate::sql_types::Foldable;
+    sql_function! {
+        #[aggregate]
+        fn my_sum<ST: Foldable>(expr: ST) -> ST::Sum;
+    }
+
+    struct MySum<T: Foldable> {
+        sum: T::Sum,
+    }
+    impl<T: Foldable> MySum<T> {
+        fn new() -> Self {
+            Self { sum: 0 }
+        }
+    }
+    impl<T: Foldable> Aggregator<T> for MySum<T> {
```

Comment:
The same comment about `T: Foldable` from above applies here too.

Additionally we need to talk about the design of the `Aggregator` trait a bit more. First of all I think it's a good idea to have a septate trait for this. That trait should be something similar to the `Fn` traits in stdlib, but with two functions instead of one. So it might be useful to have a look at [their documentation](https://doc.rust-lang.org/core/ops/trait.Fn.html). Trait bounds around callbacks are normally written as `F: Fn(Arg1, Arg2) -> ReturnType`, which is syntactic sugar for the unstable `Fn` trait. Using the unstable trait this would look like `F: Fn<(Arg1, Arg2), Output = ReturnType>`.
Additional we need some way to setup a new instance of the `Aggregator` for each call. I think you want to use the `new` function for that, but that won't work in generic context, as it's not possible to access functions that are not part of a certain trait. So just let's use `Default` for this.

That leads to the following design of the `Aggregator` trait:
```rust
trait Aggregator<Args>: Default {
    type Output;

    fn step(&mut self, args: Args);
    fn finalize(self) -> Self::Output;
}
```
where I'm not sure if `finalize` should really consume `self` (It seems to be the most reasonable choice.)

Reply:
That makes a lot of sense. Thanks for the insight!